### PR TITLE
Add parallel MsBuild support (/m) when building the generated solution.

### LINKED ISSUE #4654: RNW msbuild projects are not set up for multiproc builds
https://github.com/microsoft/react-native-windows/blob/ff6f8f57587f7364143a701ef6637039b6643cf4/vnext/local-cli/runWindows/utils/msbuildtools.js#L56

We need to survey our projects to correct the inter-project dependencies so that we can enable multi proc builds

### PR DESCRIPTION
**Pull request by dannyvv**
This fixes #4654 by adding [`/maxCpuCount`](https://docs.microsoft.com/en-us/visualstudio/msbuild/msbuild-command-line-reference?view=vs-2019) through [msbuildtools.js](https://github.com/microsoft/react-native-windows/blob/master/vnext/local-cli/runWindows/utils/msbuildtools.js)
The bug is worried about missing dependencies but from testing it seems fine. And unfortunately there is almost no perf-gain either.

# How Tested:
I tested on a 48-thread machine (2 x E5-2687W) with 64Gb ram and NVME SSD using:
1. `npx react-native init projectName --version ^0.62.2`
1. `cd projectName`
1. `npx react-native-windows-init --overwrite`
1. `npx react-native run-windows`
1. Close all opened windows
1. Open VS 2019 Developer prompt
1. `cd windows`
1. Run the commands below and record time elapsed

### `msbuild /m /t:rebuild /p:Configuration=Debug /p:Platform=x86`
1. 00:02:51.80
1. 00:02:47.07
1. 00:02:53.30

### `msbuild /m /p:Configuration=Debug /p:Platform=x86`
1. 00:00:20.53
1. 00:00:07.21
1. 00:00:06.87

### `msbuild /m /t:rebuild /p:Configuration=Release /p:Platform=x64`
1. 00:03:29.60
1. 00:03:34.83
1. 00:03:25.49

### `msbuild /m /p:Configuration=Release /p:Platform=x64`
1. 00:00:51.54
1. 00:00:40.18
1. 00:00:40.31

### `msbuild /t:rebuild /p:Configuration=Debug /p:Platform=x86`
1. 00:02:53.85
1. 00:03:05.64
1. 00:03:01.31

### `msbuild /p:Configuration=Debug /p:Platform=x86`
1. 00:00:22.00
1. 00:00:07.94
1. 00:00:07.71

# Results
I've run various builds (not keeping track) and I did not run into any ordering issues so far.
My machine only reached 100% CPU for a few seconds during the build when many cpp files were compiled, so dependencies seem fine as even with `/m` the concurrency level is holding back compile times.

This also shows as there is no significant time improvement over not passing `/m` in msbuild :(

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/react-native-windows/pull/4739)

### CODE REVIEW

**Comment by NickGerleman** [comment]
@dannyvv you can fix your CI issue by running `yarn lint:fix`. You can configure vs code to use eslint for formatting if you plan to spend time in our JS.

**Comment by dannyvv** [comment]
> @dannyvv Danny van Velzen FTE you can fix your CI issue by running `yarn lint:fix`. You can configure vs code to use eslint for formatting if you plan to spend time in our JS.

Thanks, kept running `yarn lint --fix` per the error output. In the end I edited the js in notepad as somehow my vscode is not picking up the proper eslint settings when formatting on save....

**Comment by NickGerleman** [comment]
@dannyvv would you mind fixing the error message in the script? Didn't realize we said something incorrect there.

**Comment by NickGerleman** [comment]
But yeah, you have to install the eslint plugin and do some configuration to get VSCode to do the right thing. I think it will otherwise use prettier defaults.

**Comment by asklar** [comment]
> This also shows as there is no significant time improvement over not passing /m in msbuild :(
That's a little surprising (and sad :( ). I'd like to understand this better. Do you have more data?
If you haven't yet, please take a look at the timeline view in the binlog to better understand where the time is being spent.

File: vnext/local-cli/generator-windows/index.js [hunk 1]
```diff
@@ -99,6 +99,7 @@ function copyProjectTemplateAndReplace(
     { from: path.join(srcRootPath, '_gitignore'), to: path.join(windowsDir, '.gitignore') },
     { from: path.join(srcRootPath, 'b_gitignore'), to: path.join(windowsDir, newProjectName, '.gitignore') },
     { from: path.join(srcRootPath, 'index.windows.bundle'), to: path.join(windowsDir, newProjectName, bundleDir, 'index.windows.bundle') },
+    { from: path.join(srcRootPath, 'Directory.Build.rsp'), to: path.join(windowsDir, 'Directory.Build.rsp') },
```
[the remark this answers is not in the document]
Comment (dannyvv) [reply vnext/local-cli/generator-windows/index.js:102]: Thanks, I'll put it back in the js file... I was just thinking of building the son from commandline..

**Comment by dannyvv** [comment]
> That's a little surprising (and sad :( ). I'd like to understand this better. Do you have more data?
> If you haven't yet, please take a look at the timeline view in the binlog to better understand where the time is being spent.

The numbers were emperical, I didn't record at the time but they were in the same ballpack. I will rerun and update the table in the description with numbers before.

Note MsBuild's `/m` parellism is only at the project level, so only if there is project level parellalism thsi will give benefit. It is not like more modern build engines that focus on processes level parellalism like [BuildXL](https://github.com/microsoft/buildxl), [Bazel](https://github.com/bazelbuild/bazel), [Buck](https://github.com/facebook/buck) and [Ninja](https://github.com/ninja-build/ninja). 
If msbuild has two projects with a dependency, both do codegen, compile, and link but one project has a dependency on the other, those 6 build steps would run sequential in msbuild, where at the process level, only the link might need to wait on the link of the dependent project. so the codegen and compiles could run in parralel. 
I'll try to produce a report from the msbuild binary log to show this.

File: vnext/local-cli/generator-windows/index.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (dannyvv) [reply vnext/local-cli/generator-windows/index.js:102]: Note the new [Sdk-Style ](https://docs.microsoft.com/en-us/visualstudio/msbuild/how-to-use-project-sdk?view=vs-2019) projects have this a lot less, even though the sln file format is still ugly.

**Comment by NickGerleman** [comment]
@asklar are you still waiting on anything?

**Comment by asklar** [comment]
> > That's a little surprising (and sad :( ). I'd like to understand this better. Do you have more data?
> > If you haven't yet, please take a look at the timeline view in the binlog to better understand where the time is being spent.
> 
> The numbers were emperical, I didn't record at the time but they were in the same ballpack. I ~will rerun and update~ reran and updated the table in the description with numbers before.
> 
> Note MsBuild's `/m` parellism is only at the project level, so only if there is project level parellalism thsi will give benefit. It is not like more modern build engines that focus on processes level parellalism like [BuildXL](https://github.com/microsoft/buildxl), [Bazel](https://github.com/bazelbuild/bazel), [Buck](https://github.com/facebook/buck) and [Ninja](https://github.com/ninja-build/ninja).
> If msbuild has two projects with a dependency, both do codegen, compile, and link but one project has a dependency on the other, those 6 build steps would run sequential in msbuild, where at the process level, only the link might need to wait on the link of the dependent project. so the codegen and compiles could run in parralel.
> ~I'll try to produce a report from the msbuild binary log to show this.~
> 
> ## The timeline view of the binlog
> image
> This demonstrates that there are 5 projects loaded in parralel, but due to dependencies on one another they are blocking.
> The dependecnies are:
> 
> * sln
>   
>   * main project
>     
>     * Microsoft.ReactNative
>       
>       * JSI.Universal
>         
>         * Common
>       * ReactCommon
>         
>         * Folly
>       * ReactWindowsCore
>         
>         * Common
>         * Folly
>         * ReactCommon
>       * Common
>       * Folly
> 
> So only the pairs (Folly, Common) and (Jsi.Universal,ReactWindowsCore) can technically run in parallel, only folly and common run parralell in practice, but given they are pretty small there is not much gain.
> So we end up with a critical path of:
> 
> ```
> sln -> main project -> Microsoft.ReactNative -> ReactWindowsCore -> ReactCommon -> Folly
> ```
> 
> [msbuild.binglog.zip](https://github.com/microsoft/react-native-windows/files/4553025/msbuild.binglog.zip)
> 
> **Update:**
> Some more breakdown of costs:
> 
> ### CL (1m 53s, 16 invocations)
> ```
>     Most have reasonable compile times as they parellalize inside cl.exe via /MP.
>     There is one that stands out and is the pch compilation...
> It takes about 30 seconds. The pre processed file is 28 megs so not horrible, yet breaking down the costs in the compiler:
> 	   612 files     -> 17.2 seconds
> 	82,633 classes   ->  7.7 seconds
> 	49,867 functions ->  8.8 seconds
> It is mostly Xaml winrt generated code
> ```
> 
> ### Midl (21.5 s)
> in Microsoft.ReactNative Midl runs 22 times sequentially totaling 19 seconds
> if that could be paralyzed it could save 17 seconds. Note `/m` has no effect here as it is in a single project. The task implementation must code parallelism itself. the cpp targets just use item expansion here, so no multi-threading.
> 
> ### mdmerge (12 s)
> called twice each taking 6-7 seconds
> 
> ### cppwinrt (9 s)
> Called 6 times (each project generates for the three parts: _plat, _ref and _comp)
> _plat is a total waste since it is done for both project generating identical files from the windows sdk...
> 
> ### mspdbcmf (11.5 s)
> Did not dive into this tool yet

thanks, love seeing this data @dannyvv !

**Pull request merged**